lib/std/crypto/ecdsa.zig

```diff
@@ -91,15 +77,15 @@ pub fn Ecdsa(comptime Curve: type, comptime Hash: type) type {
             s: Curve.scalar.CompressedScalar,
 
             /// Create a Verifier for incremental verification of a signature.
-            pub fn verifier(self: Signature, public_key: PublicKey) (NonCanonicalError || EncodingError || IdentityElementError)!Verifier {
-                return Verifier.init(self, public_key);
+            pub fn verifier(self: Signature, comptime Hash: type, public_key: PublicKey) (NonCanonicalError || EncodingError || IdentityElementError)!Verifier {
+                return Verifier(Hash).init(self, public_key);
             }
 
             /// Verify the signature against a message and public key.
             /// Return IdentityElement or NonCanonical if the public key or signature are not in the expected range,
             /// or SignatureVerificationError if the signature is invalid for the given message and key.
-            pub fn verify(self: Signature, msg: []const u8, public_key: PublicKey) (IdentityElementError || NonCanonicalError || SignatureVerificationError)!void {
-                var st = try Verifier.init(self, public_key);
+            pub fn verify(self: Signature, comptime Hash: type, msg: []const u8, public_key: PublicKey) (IdentityElementError || NonCanonicalError || SignatureVerificationError)!void {
```

Comment:
Now having to pass the hash function to every function is very unusual, and bug-prone. This is also inconsistent with EdDSA. And is yet another breaking change that may annoy users.

To avoid lists of types for all pairs, wouldn't something like that work?

```zig
fn EcdsaBuilder(comptime Curve: type) {
    return struct {
       fn Build(comptime Hash: type) type {
          return Ecdsa(Curve, Hash);
       }
   };
}
```

If we want to avoid inline switches, and do actual dynamic dispatch, we can also replace `Hash` with a pointer to a generic IUF interface implemented by all hash functions.

Reply:
Ah, you're correct. I thought you could pair any ECDSA curve with any SHA2 function, but [Mozilla says:](https://www.mozilla.org/en-US/about/governance/policies/security-group/certs/policy/#512-ecdsa)
> - If the signing key is P-256, the signature MUST use ECDSA with SHA-256
> - If the signing key is P-384, the signature MUST use ECDSA with SHA-384.

I'll revert to what was there previously and enforce that.

There's still a need to parameterize the hash function for RSA since any modulus length can use any SHA2 function. I think dynamic dispatch would be better for code size but worse for comptime security checks. We can measure the tradeoff in the future.